Modify `get_movie` to display frames as they are collected

Baharis opened the pull request:
### Context
As mentioned in my previous therads (#111, #116, #118), instamatic's `get_movie` protocol, while implemented a long time ago, is yet to see use in instamatic. Compared to `get_image`, it has more drawbacks than benefits: it is not faster, it does not collect header data, and it does not display images as soon as they are collected. This PR addresses this issues.

Firstly, as suggested in issue #118, for all defined cameras the `cam.get_movie` method has been rewritten to return an instance of np.ndarray *generator* instead of a np.ndarray *list*:
```python
In [1]: type(ctrl.cam.get_image(exposure=0.1))
Out[1]: numpy.ndarray

In [2]: type(ctrl.cam.get_movie(n_frames=1, exposure=0.1))
Out[2]: generator
```

The generator `yields` new output in-the-fly. In particular, calling `get_movie` does NOT start a movie; rather, it creates a movie generator that runs collection only when it is iterated i.e. by calling `next(gen)` or using it as an iterable in a loop:
```python
In [3]: for frame in ctrl.cam.get_movie(n_frames=1, exposure=0.1)): print(type(frame))
<class 'numpy.ndarray'>
```

While this adds a level of complexity compared to a simple image list, new frames are yielded as soon as they are collected rather than only when the entire function finishes. New `VideoStream`  uses this fact to dynamically update the displayed `frame` in the GUI as soon as it is collected and yield it to `TEMController`, which in turn can immediately enhance it with metadata:
```python
from datetime import datetime
for i, h in ctrl.get_movie(n_frames=5, exposure=2.0): 
    print(f'{h["ImageGetTimeEnd"]}: new {i.shape}-shape image collected.')
    
2174435.4982932: new (516, 516)-shape image collected.
2174437.4999097: new (516, 516)-shape image collected.
2174439.5015133: new (516, 516)-shape image collected.
2174441.5034781: new (516, 516)-shape image collected.
2174443.5368059: new (516, 516)-shape image collected.
```

The input parameters of all `cam.get_movie` remain unchanged but the signature of `ctrl.get_movie` was adapted to accept new parameters, including `comment` and headers handled already by `get_image`. In addition to standard `header_keys` that, similarly to `get_image`, collects metadata for each frame, a new `header_keys_common` specifies header fields that should be collected only once because they do not change during the measurement. The "variable" and common headers are specified in `TEMController.MOVIE_HEADER_KEYS_VARIABLE` and `TEMController.MOVIE_HEADER_KEYS_COMMON`. Still, if you are not interested in headers and would just prefer `get_movie` to return you a simple list of `np.ndarray` as before, you can put the results in a list:
```python
old_movie = [i for i, _ in ctrl.get_movie(n_frames=5, exposure=2.0, header_keys=(), header_keys_common=())]
```

If you want to check out the new get_movie displaying images "in action" i.e. in a live GUI, run the GUI and from the debug tab execute a newly-included `instamatic/config/scripts/showcase_movie.py`. This script simply collects the movie, which due to the introduced changes is dynamically displayed in the stream frame. It also compares the timing of `get_movie`, a for-loop of `get_image` (should be slightly slower) and a for-loop of `tqdm(get_image)` which, as I discovered, becomes extremely slow when using GUI, most likely because it interacts poorly with the console frame.

![image](https://github.com/user-attachments/assets/fa9ce47f-79b3-4bd7-b076-b2e5e5935bc2)

The script is not intended as a permanent addition and can be removed before merge.

### Calibration
Neither previous nor current version of `get_movie` does not offer precise temporal spacing of frames. This is because every detector differs in terms of initial delay, dead time, or ultimate return time. I am of an opinion that every microscope user should be more of less aware of these parameters when planning their experiments. However, if an user would like to estimate the initialization or readout of their detector, this PR additionally introduce a new script called `instamatic.calibrate_movie_delays`.

Imagine that you (have a way to) schedule a cRED experiment with goniometer speed of 1 degree per second and movie of 1-second exposure. After collecting the data you process it and realize that the corrected mean distance between your frames is, in fact, 1.01 degree. Why is that? Presumably your detector has a 10ms readout time between `get_movie` frame. So by setting exposure of 1 second, you get a new image every 1.01 second. This calibration is again designed to show you that and give you an option to correct for the added dead time.

In the new cRED frame that I am preparing, it will be most likely up to the user to input their dead time to correct for it; I am not certain the calibration will be called directly to estimate the delay automatically. Nevertheless I decided to include the calibrate script in this PR so that a user has an option to check their delay and make an informed guess about their equipment parameters, should they not know it. The new calibration script collects five individual 20-frame movies with desired exposure and headers, displays the results, and saves a calibration object to a new file `%appdata%/instamatic/calib/calib_movie_delays.json`. I personally find it a good place for this information but it is something to discuss.

![image](https://github.com/user-attachments/assets/b9792484-344b-47c0-998d-920741cc902f)

### Changes

- All `get_movie` methods are now generator functions that yield images as soon as they are collected 
- An `instamatic.calibrate_movie_delays` was added to estimate microscope initialization, return, dead-time delays.
- In `CameraServal`, use `tifffile.imread()` instead of `np.array(PIL.Image())` to cut read times by half;
- In `TEMController`, rewrite a cam attribute check into a neat decorator

### Bugfixes
- `ctrl.get_image`, `get_movie`: following doc-strings now correctly accept tuples of headers, not single header string.

### Effects on the database
With this feature, `get_movie` can be used by instamatic frames to collect data for long scans, cRED etc. whilst displaying the data in the GUI window all the time. As mentioned in #118 and shown on the top image here, for the simulated camera the init + dead time between 100 frames accumulates in total to 200ms, which can be attributed mostly to the wait time between individual frames. Coincidentally, this wait time is very similar for our Cheetah detector, though the initialization is much slower.

stefsmeets reviewed (changes requested):
Looks good to me for the most part. The PR is a quite long, so I'm a bit hesitant with it. The main bottleneck for me is the complexity of the calibrate_movie_delays.py function. It's very difficult for me to gauge what is happeneing there.

I also wonder whether json is the best fileformat to store these data. I'd prefer to use yaml (to be in line with the rest of the code). Also, I wonder if a tabular data format would not be a better choice.

Baharis commented on src/instamatic/calibrate/calibrate_movie_delays.py:
```diff
@@ -0,0 +1,402 @@
+    h = 'Exposure to test the delay for in seconds. Default: 1'
+    parser.add_argument('-e', '--exposure', type=float, default=1.0, help=h)
+
+    h = 'Comma-delimited list of variable header keys to calibrate for. '
+    h += 'For default, see `src/instamatic/controller.py:MOVIE_HEADER_KEYS_VARIABLE`.'
+    parser.add_argument('-a', '--variable_headers', type=str, default=None, help=h)
```
I personally much prefer this dense style (as you can probably gauge from the rest of the commit) because with that the entire `main_entry` easily fits on 1 screen but I can convert these for the sake of consistency with the rest of code base.

Baharis commented on src/instamatic/camera/camera_serval.py:
```diff
@@ -123,27 +112,43 @@ def _get_image_single(self, exposure: float, **_) -> np.ndarray:
-        img = self.conn.get_image_stream(nTriggers=1, disable_tqdm=True)[0]
-        arr = np.array(img)
-        return arr
+        response = self.conn.get_request('/measurement/image')
+        return tifffile.imread(BytesIO(response.content))
```
Serval can transfer pgm, png, tiff, jsonimage, and some other; the first three are natively supported by ASI-provided serval toolkit for python. Tiff are the by far the fastest format in terms of total serializing + transfer + de-serializing time. Thus, I've modified `ServalCamera` to always use tiff. This was they key point of #111 .

To summarize the discussion there, both pgm and png formats take PIL longer to de-serialize than tiff. Tiff files are indeed slightly larger to transfer (~8 ms transfer for our 514 x 514 detector) but it allows > 16-bit images and does not get extremely confused by non 2**N-bit rates of pgm that might be faster to transfer, but take PIL ages to open:

```
file_format                pgm             png            tiff
bit_depth                                                    
1             117.41 ± 6.32 ms   nan ±  nan ms  0.17 ± 0.03 ms
4            117.24 ± 11.10 ms   nan ±  nan ms  0.17 ± 0.04 ms
6            151.23 ± 48.35 ms   nan ±  nan ms  0.16 ± 0.01 ms
8              0.08 ±  0.22 ms  1.41 ± 0.09 ms  0.20 ± 0.07 ms
10           153.95 ± 13.69 ms   nan ±  nan ms  0.63 ± 0.10 ms
12           153.15 ± 23.07 ms   nan ±  nan ms  0.46 ± 0.17 ms
14           150.50 ± 17.25 ms   nan ±  nan ms  0.68 ± 0.11 ms
16             1.84 ±  0.20 ms  2.10 ± 0.15 ms  0.42 ± 0.15 ms
20              nan ±   nan ms   nan ±  nan ms  2.17 ± 0.22 ms
24              nan ±   nan ms   nan ±  nan ms  2.21 ± 0.21 ms
32              nan ±   nan ms   nan ±  nan ms  2.22 ± 0.19 ms
```

As for why I am suggesting tifffile here instead of PIL, I simply realized that it is faster. I modified the benchmark used to compare formats before to compare PIL with tifffile instead and got the following:
```
method                PIL        tifffile
bit_depth                                
1          0.31 ± 0.96 ms  0.27 ± 0.05 ms
4          0.28 ± 0.04 ms  0.28 ± 0.06 ms
6          0.30 ± 0.06 ms  0.27 ± 0.05 ms
8          0.29 ± 0.05 ms  0.25 ± 0.06 ms
10         0.36 ± 0.04 ms  0.31 ± 0.07 ms
12         0.38 ± 0.08 ms  0.30 ± 0.07 ms
14         0.36 ± 0.06 ms  0.28 ± 0.05 ms
16         0.38 ± 0.09 ms  0.29 ± 0.07 ms
20         4.19 ± 0.39 ms  0.88 ± 0.14 ms
24         3.95 ± 0.35 ms  0.81 ± 0.15 ms
32         4.04 ± 0.38 ms  0.83 ± 0.16 ms
```

This is not broken and if you think this deserves its own space, I can remove it from this PR, but I thought to include the accidental improvement that I found. Personally, I am more surprised about your question about tiff inefficiency? I am under the impression it is as good as it goes without writing custom de/serialisers for 12- or 24-bit images, which is something I might eventually think about but did not want to be bothered by in #111.

stefsmeets replied:
Tiff is a pretty overloaded and complex format 😅 I was under the impression we were transferring byte arrays before, which is why this feels like a step backwards. That said, your timings are convincing enough.

Baharis replied:
IIRC Serval allows transfering jsonimage, which is tiny JSON with a binary bit array. If the transfer/conversion speeds are ever an issue, I can work on using these.

Baharis commented on src/instamatic/camera/camera_serval.py:
```diff
@@ -123,27 +112,43 @@ def _get_image_single(self, exposure: float, **_) -> np.ndarray:
+            self.conn.measurement_start()
+            for i in range(n_frames):
+                response = self.conn.get_request('/measurement/image')
+                yield tifffile.imread(BytesIO(response.content))
```
Answered in the other thread.

Baharis commented:
A accepted some suggestions and merged others. This code underwent a lot of modifications and hence it has some really rough edges, like methods `str_to_ehh` and `eeh_to_str` that indeed are hilariously opaque and need a second look. I would argue it is not particularly long; *numerically* most of the additions are in the calibration file, which itself is shorter than its counterparts.

I do not mind switching json to yaml. Or csv, as this data is absolutely tabular, though this will make it less human-readable due to the keys being variable-sized.

As for what the main part of the calibration is doing, code should be self-explanatory so I am going to rewrite it, fix doc-string, improve names etc. However, if you don't want to wait for that: It take exposure time, header keys, and common header keys. For a combination of these, it collects movie a few times and times the output. Based on that it estimates how long it takes for movie to start (init_time), finish (return_time), what are the delays between frames (dead_time). This information is then used when e.g. planning a cRED experiment with controlled speed: if you want 1 frame per 1 deg, you set exposure to 1 sec, measure delay - say, 10 ms - and slow the rotation speed by 1% to adjust for the dead time between frames. Similarly you can try to use init or return time, but I found these less reliable.

Baharis commented:
Alright, I have worked on this PR a bit more today and changed the following:
- the "section comments" (# ~~~~ #) are gone;
- improved the readability of `CalibMovieDelaysMapping` methods;
- basically rewritten all major doc-strings from scratch (previously some were inaccurate and others straight-up wrong);
- replaced json with yaml, IMO the keys look much worse but is consistent (see below);
- completely refactored, shortened, improved readability of `calibrate_movie_delays_live`;
Also, a note: this PR is pretty large, currently +708 -104, but 468 of that is a new calibration, ~30 links to this calibration, ~40 a showcase script, so in fact it adds only ~60 and replaces ~100 lines of the existing code.

Json format:
```json
{
    "0.000;;": {
        "init_time": 0.0286700241512766,
        "return_time": 3.5599805414676666e-06,
        "wait_time": 0.002286327904239956,
        "yield_time": 6.83309929445386e-05
    },
    "0.100;;": {
        "init_time": 0.02208434951148535,
        "return_time": 4.220008850097656e-06,
        "wait_time": 0.0034395147114992114,
        "yield_time": 8.787202183157205e-05
    },
    "0.100;BeamShift,BeamTilt,StagePosition;Brightness,DiffFocus,DiffShift,FunctionMode,GunShift,GunTilt,ImageShift1,ImageShift2,Magnification,SpotSize": {
        "init_time": 0.027397886291146277,
        "return_time": 3.660004585981369e-06,
        "wait_time": 0.003689832627577208,
        "yield_time": 9.138700552284718e-05
    },
}
```

Yaml format:
```yaml
0.000;;:
  init_time: 0.0286700241512766
  return_time: 3.5599805414676666e-06
  wait_time: 0.002286327904239956
  yield_time: 6.83309929445386e-05
? 0.010;BeamShift,BeamTilt,StagePosition;Brightness,DiffFocus,DiffShift,FunctionMode,GunShift,GunTilt,ImageShift1,ImageShift2,Magnification,SpotSize
: init_time: 0.025899659475544467
  return_time: 2.2655002248939125e-05
  wait_time: 0.0029916402632309317
  yield_time: 8.352374999958555e-05
0.100;;:
  init_time: 0.028958414737012626
  return_time: 2.1379999816417693e-05
  wait_time: 0.0038562536833954887
  yield_time: 9.516500082099811e-05
? 0.100;BeamShift,BeamTilt,StagePosition;Brightness,DiffFocus,DiffShift,FunctionMode,GunShift,GunTilt,ImageShift1,ImageShift2,Magnification,SpotSize
: init_time: 0.027397886291146277
  return_time: 3.660004585981369e-06
  wait_time: 0.003689832627577208
  yield_time: 9.138700552284718e-05
0.500;;:
  init_time: 0.026543468371719947
  return_time: 8.199946023523808e-06
  wait_time: 0.0037491631902459854
  yield_time: 9.322997066192329e-05
1.000;:
  init_time: 0.027216010414505054
  return_time: 1.659989356994629e-05
  wait_time: 0.005775965780853043
  yield_time: 0.00019534751190803946
```

stefsmeets commented:
Looks fine to me for the most part. My main issue is that the format of the calibration file does not make sense to me. It is difficult to read. The main thing for me are the mangled key names with both the timings and the keys in there. I think it would be beneficial to think about how you want to represent the data in the output, e.g. something like this:

```yaml
- exposure_time: 0.000
  init_time: 0.0286700241512766
  return_time: 3.5599805414676666e-06
  wait_time: 0.002286327904239956
  yield_time: 6.83309929445386e-05
  common_header_keys: []
  variable_header_keys: []
- exposure_time: 0.010
  init_time: 0.025899659475544467
  return_time: 2.2655002248939125e-05
  wait_time: 0.0029916402632309317
  yield_time: 8.352374999958555e-05
  common_header_keys: [BeamShift, BeamTilt, StagePosition]
  variable_header_keys: [Brightness, DiffFocus, DiffShift, FunctionMode, GunShift, GunTilt, ImageShift1, ImageShift2, Magnification, SpotSize]
- exposure_time: 0.100
  init_time: 0.028958414737012626
  return_time: 2.1379999816417693e-05
  wait_time: 0.0038562536833954887
  yield_time: 9.516500082099811e-05
  common_header_keys: []
  variable_header_keys: []
```

I think this will be far easier to parse, and you won't need the complex conversion methods for the dictionary keys. Let me know what you think.

Baharis commented:
I think this representation of the calibration file is *nicer to look at* but it represents a misleading information. My format represents a mapping: if A then B i.e. if exposure & headers then times. The *nice* format presents a list: there are some A & B i.e. some exposures, headers & times. It is unfortunate that neither yaml nor json allows complex keys, but at the same time I have an impression we are discussing something extremely minor. Personally I did not give it much attention because none of the existing calibration files are readable at all, but I can make it look nicer.

Baharis commented:
I have updated the `CalibMovieDelaysMapping` class responsible for de/serializing calibration results to read/write them from/to a really freaking adorable yaml file that should now be absolutely unnecessarily well-readable, especially in comparison to the rest of the pickle kingdom of `instamatic.calibrate`. Hope that addresses your last concerns:

```yaml
- calibration_conditions:
    exposure: 0.1
    header_keys_common: []
    header_keys_variable: []
  calibration_results:
    init_time: 0.029504721052772333
    return_time: 1.5959999655024148e-05
    wait_time: 0.0035669294735845626
    yield_time: 8.644700010336237e-05
- calibration_conditions:
    exposure: 0.3
    header_keys_common:
    - BeamShift
    - BeamTilt
    - StagePosition
    header_keys_variable:
    - Brightness
    - DiffFocus
    - DiffShift
    - FunctionMode
    - GunShift
    - GunTilt
    - ImageShift1
    - ImageShift2
    - Magnification
    - SpotSize
  calibration_results:
    init_time: 0.009939670526249467
    return_time: 1.5660000099160242e-05
    wait_time: 0.002698230526278311
    yield_time: 0.00010350400004426775
- calibration_conditions:
    exposure: 0.01
    header_keys_common:
    - FunctionMode
    - GunShift
    - GunTilt
    - ImageShift1
    - ImageShift2
    - DiffShift
    - Magnification
    - DiffFocus
    - Brightness
    - SpotSize
    header_keys_variable:
    - BeamShift
    - StagePosition
  calibration_results:
    init_time: 0.029696983157765845
    return_time: 1.6340000183845404e-05
    wait_time: 0.003250398947380553
    yield_time: 8.030799999687589e-05
- calibration_conditions:
    exposure: 0.01
    header_keys_common: []
    header_keys_variable:
    - BeamShift
    - StagePosition
  calibration_results:
    init_time: 0.017882577894746836
    return_time: 1.9240000074205455e-05
    wait_time: 0.0029236831579634638
    yield_time: 7.270399992194143e-05
```

stefsmeets commented:
> to a really freaking adorable yaml file that should now be absolutely unnecessarily well-readable, especially in comparison to the rest of the pickle kingdom of `instamatic.calibrate`

Thanks, much appreciated! As ruler of pickle kingdom, I completely agree with you that this was not the best choice in hindsight. 😅
